engine/crates/engine-v2/src/execution/coordinator.rs

```diff
@@ -350,90 +353,153 @@ impl SubscriptionResponse {
-    futures: ExecutionPlanFutureSet<'exec, <R::Hooks as Hooks>::OnSubgraphResponseOutput>,
     executed_operation_builder: ExecutedOperationBuilder<<R::Hooks as Hooks>::OnSubgraphResponseOutput>,
     state: OperationExecutionState<'ctx>,
     response: ResponseBuilder,
 }
 
-impl<'ctx, 'exec, R: Runtime> std::ops::Deref for OperationExecution<'ctx, 'exec, R> {
+impl<'ctx, R: Runtime> std::ops::Deref for OperationExecution<'ctx, R> {
     type Target = ExecutionContext<'ctx, R>;
     fn deref(&self) -> &Self::Target {
         &self.ctx
     }
 }
 
-impl<'ctx, 'exec, R: Runtime> OperationExecution<'ctx, 'exec, R>
-where
-    'ctx: 'exec,
-{
+enum State<F, S> {
+    Ingestion(F),
+    Execution(S),
+}
+
+impl<'ctx, R: Runtime> OperationExecution<'ctx, R> {
     /// Runs a single execution to completion, returning its response
-    async fn run(mut self) -> Response<<R::Hooks as Hooks>::OnOperationResponseOutput> {
+    async fn run(
+        mut self,
+        mut results: VecDeque<ExecutionPlanResult<<R::Hooks as Hooks>::OnSubgraphResponseOutput>>,
+    ) -> Response<<R::Hooks as Hooks>::OnOperationResponseOutput> {
+        let futures = FuturesUnordered::new();
         for plan_id in self.state.get_executable_plans() {
-            self.spawn_resolver(plan_id);
+            if let Some(fut) = self.create_plan_execution_future(plan_id) {
+                futures.push(fut);
+            }
+        }
+
+        let mut state = State::Execution(self);
+
+        futures_util::pin_mut!(futures);
+        let self_ = loop {
+            state = match state {
+                State::Ingestion(mut ingestion_fut) => {
+                    let next_task = futures_util::select_biased! {
+                        ingestion_result = ingestion_fut => Ok(ingestion_result),
+                        execution_result = futures.next() => Err(execution_result),
+                    };
+                    match next_task {
+                        Ok((self_, next_futures)) => {
+                            for fut in next_futures {
+                                futures.push(fut);
+                            }
+                            State::Execution(self_)
+                        }
+                        Err(Some(result)) => {
+                            results.push_back(result);
+                            State::Ingestion(ingestion_fut)
+                        }
+                        Err(None) => {
+                            let (self_, next_futures) = ingestion_fut.await;
+                            for fut in next_futures {
+                                futures.push(fut)
+                            }
+                            State::Execution(self_)
+                        }
+                    }
+                }
+                State::Execution(self_) => {
+                    if let Some(result) = results.pop_front() {
+                        State::Ingestion(Box::pin(self_.ingest_execution_result(result).boxed().fuse()))
```

Comment:
Looks like we're boxing this twice?  Do we need to do that?

Reply:
oh yes my bad, definitely not